hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/services/ListBlobProducer.java

```diff
@@ -0,0 +1,70 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.hadoop.fs.azurebfs.services;
+
+import org.apache.hadoop.fs.azurebfs.contracts.exceptions.AzureBlobFileSystemException;
+import org.apache.hadoop.fs.azurebfs.utils.TracingContext;
+
+public class ListBlobProducer {
+
+  private final AbfsClient client;
+
+  private final ListBlobQueue listBlobQueue;
+
+  private final String src;
+
+  private final TracingContext tracingContext;
+
+  private String nextMarker;
+
+  public ListBlobProducer(final String src,
+      final AbfsClient abfsClient,
+      final ListBlobQueue listBlobQueue,
+      final String initNextMarker,
+      TracingContext tracingContext) {
+    this.src = src;
+    this.client = abfsClient;
+    this.tracingContext = tracingContext;
+    this.listBlobQueue = listBlobQueue;
+    listBlobQueue.setProducer(this);
+    this.nextMarker = initNextMarker;
+    new Thread(() -> {
```

Comment:
Should we make use of ExecutorService instead of a creating a new thread manually ?

Reply:
ExecutorService will constraint the number of producers running in parallel. If there are n threads calling rename API. We should be able to spawn n producers in parallel. Else, producer spawning will become sequential. Hence, spawning thread manually here.

What kind of interruption can happen? There is no external scope which can interrupt the thread.

Nevertheless, I will add handling if interruption happen, the same gets propogated to the queue which stops the consumer.

Reply:
There should always be a cap on number of threads which can be spawned in parallel. This number n can be a very huge number as well and can lead to resource exhaustion and starvation

Reply:
Is there a cap on the n threads which can call the rename API ?

Reply:
There can be n number of threads that dev spawn which can hit rename API on the FileSystem class. The dev will be 
 and has to be aware that this will have context-switching.
The problem if we dont create n number of producers is:
lets say we can create only m producers, where m < n.
So, the 'm' dev threads will be able to have the renaming working for it. But, 'n-m' dev threads would be stalled and do nothing. The bad part is that those 'n-m' threads would get an equal share of CPU where it would be just waiting on queue to be filled. So, we would lose onto the time received by 'n-m' threads.

Reply:
For the interruption,
first there is no other thread that is going to call interrupt on the producer thread.
Second, interrupt () method does not terminate the thread. It only sends a signal to the thread to stop what it is doing and do something else. An interrupt can occur at any time during the execution of a thread, and it does not depend on the scheduling policy of the operating system. An interrupt can be caused by hardware devices, software exceptions, or user actions. The thread resumes back on what it was doing after the interrupt handler returns. The interrupt handler is a special routine that runs in response to an interrupt and performs some urgent task.

ex: 
```
public void test() throws Exception {
    Thread thread = new Thread(() -> {
      for(int i=0;i<10;i++) {
        try {
          System.out.println("iter " + i);
          Thread.sleep(1000);
        } catch (InterruptedException e) {
          System.out.println("Interrupt");
        }
      }
      System.out.println("completed!");
    });

    thread.start();

    Thread.sleep(5000);
    thread.interrupt();

    Thread.sleep(10000);
  }
```

gives:
```
iter 0
iter 1
iter 2
iter 3
iter 4
Interrupt
iter 5
iter 6
iter 7
iter 8
iter 9
completed!
````

Reply:
There can be n number of threads that dev spawn which can hit rename API on the FileSystem class.  
So are you suggesting that the rename API can be called in parallel with any number of threads without any cap ?

Reply:
yes, app using hadoop-azure is free to spawn as many threads they want.